### ISSUE TITLE
ctrl+c causes bats to hang

### ISSUE THREAD

**Issue by kaysond** [bats-core/bats-core#353]
Running bats v1.2.1 on ubuntu 20.04. If I want to kill a run early, and I hit ctrl+c, the process just gets stuck. I end up having to ctrl+z to put in the background, then find the bash process and kill it.

Is there another interrupt that should be used to kill it mid-run?

**Comment by martin-schulze-vireso** [comment]
Thanks for the report. I saw similar behavior when using `--jobs` > 2. Do you have a reliable reproducer (file + bats call)?

**Comment by kaysond** [comment]
Oops I lied. I'm not actually using the v1.2.1 release. I'm using HEAD.

But in any case, if you `git clone https://github.com/kaysond/spki.git`, then run `bats test/test.bats`, it always hangs on `ctrl+c`

**Comment by martin-schulze-vireso** [comment]
So just to be sure, with run you mean calling `bats ./test/test.bats`, no `--jobs` no other flags?

**Comment by kaysond** [comment]
> So just to be sure, with run you mean calling `bats ./test/test.bats`, no `--jobs`, no other flags?

Correct. No flags whatsoever

**Comment by martin-schulze-vireso** [comment]
Okay, I can reproduce this within docker (ubuntu:20.04).

**Comment by martin-schulze-vireso** [comment]
I think I found the culprit now. It was this loop in the `bats_exit_trap` of `bats-exec-test`:
```
    while IFS= read -r line; do
      printf '# %s\n' "$line"
    done <"${BATS_OUT?}" >&3
```

When we send SIGINT via CTRL-C, the exit handler gets executed within the context of the test. The test's output is redirected into `$BATS_OUT`. I looked into this file and saw it filling up with the error: 

> /bats-core/libexec/bats-core/bats-exec-test: line 107: printf: write error: Broken pipe

So I assume that fd 3 has already gone away due to CTRL-C. The error message is put on stderr which is also redirected into `$BATS_OUT` which results in the loop talking to itself through this file. We can fix this by exiting (or breaking the loop) when `printf` returns non-zero. I don't know why this does not happen automatically, since `set -eE` is done in this file, maybe it does not work inside the trap itself.

@sublimino 
Digging through all of this made me realize that fd 3 handling is not working as promised anymore! We now have `exec 3<&1` in `bats-exec-file` and `bats-exec-test` which results in `echo Test >&3` not being printed directly to stdout but rather being buffered in `bats-exec-file` now. 

**Update**

While working on a reproducer I noticed that the problem only exists if we interrupt a command inside `run`, which starts with `set +eET`. I assume that this disabling of the error trap creates the problem in the loop.
